cmd/apigen/main.go

```diff
@@ -0,0 +1,395 @@
+		}
+		apiResourceSchema, err := readAPIResourceSchema(path, gr)
+		if err != nil {
+			return fmt.Errorf("failed to read %s: %w", info.Name(), err)
+		}
+		apiResourceSchemas[gr] = apiResourceSchema
+		logger.Info(fmt.Sprintf("Loaded APIResourceSchema for %s from %s", gr.String(), path))
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	logger.Info(fmt.Sprintf("Loaded %d APIResourceSchemas", len(apiResourceSchemas)))
+	return apiResourceSchemas, nil
+}
+
+func readAPIResourceSchema(path string, gr metav1.GroupResource) (*apisv1alpha1.APIResourceSchema, error) {
+	raw, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("could not read APIResourceSchema %s: %w", gr.String(), err)
+	}
+
+	expectedGvk := &schema.GroupVersionKind{Group: apis.GroupName, Version: "v1alpha1", Kind: "APIResourceSchema"}
+
+	obj, gvk, err := kcpscheme.Codecs.UniversalDeserializer().Decode(raw, expectedGvk, &apisv1alpha1.APIResourceSchema{})
+	if err != nil {
+		return nil, fmt.Errorf("could not decode raw APIResourceSchema %s: %w", gr.String(), err)
+	}
+
+	if !equality.Semantic.DeepEqual(gvk, expectedGvk) {
+		return nil, fmt.Errorf("decoded APIResourceSchema %s into incorrect GroupVersionKind, got %#v, wanted %#v", gr.String(), gvk, expectedGvk)
+	}
+
+	apiResourceSchema, ok := obj.(*apisv1alpha1.APIResourceSchema)
+	if !ok {
+		return nil, fmt.Errorf("decoded APIResourceSchema %s into incorrect type, got %T, wanted %T", gr.String(), obj, &apisv1alpha1.APIResourceSchema{})
+	}
+
+	return apiResourceSchema, nil
+}
+
+func convertToSchemas(prefix string, crds map[metav1.GroupResource]*apiextensionsv1.CustomResourceDefinition) (map[metav1.GroupResource]*apisv1alpha1.APIResourceSchema, error) {
+	apiResourceSchemas := map[metav1.GroupResource]*apisv1alpha1.APIResourceSchema{}
+	for gr, crd := range crds {
+		apiResourceSchema, err := apisv1alpha1.CRDToAPIResourceSchema(crd, prefix)
+		if err != nil {
+			return nil, err
+		}
+		apiResourceSchemas[gr] = apiResourceSchema
+	}
+	return apiResourceSchemas, nil
+}
+
+func resolveLatestAPIResourceSchemas(logger logr.Logger, previous, current map[metav1.GroupResource]*apisv1alpha1.APIResourceSchema) map[metav1.GroupResource]*apisv1alpha1.APIResourceSchema {
+	resolved := map[metav1.GroupResource]*apisv1alpha1.APIResourceSchema{}
+	for gr, currentSchema := range current {
+		resolvedSchema := currentSchema
+		if previousSchema, existed := previous[gr]; existed {
+			if diff := cmp.Diff(previousSchema.Spec, currentSchema.Spec, compareSchemas()); diff == "" {
+				logger.Info(fmt.Sprintf("Using previous APIResourceSchema for %s, as no changes were detected.", gr.String()))
+				resolvedSchema = previousSchema
+			}
+		}
+		resolved[gr] = resolvedSchema
+	}
+	return resolved
+}
+
+// compareSchemas compares JSON Schemas by unmarshalling them and comparing their values, instead
+// of comparing their raw []byte() representations, as those are not semantically meaningful
+func compareSchemas() cmp.Option {
```

Comment:
@ncdc this is what we were chatting about in Slack - `equality.Semantic.DeepEqual()` _et. al._ do not consider the objects equal when literally comparing `[]byte()` because the order of the fields in the JSON/YAML document changes between the in-memory version and what's written to disk.

Side-note: this is the sort of featureset that makes me partial to `cmp.Diff()` :)